nep-dapi.mediawiki

```diff
@@ -0,0 +1,586 @@
+
+'''Error return value'''
+
+string - `NO_PROVIDER`|`CONNECTION_DENIED`
+
+'''Example'''
+
+<pre>
+getAccount()
+.then((account: Account) => {
+  const {
+    address,
+    publicKey,
+  } = account;
+
+  console.log('Provider address: ' + address);
+  console.log('Provider public key: ' + publicKey);
+})
+.catch((err: string) => {
+  switch(err) {
+    case NO_PROVIDER:
+      console.log('No provider available.');
+      break;
+    case CONNECTION_DENIED:
+      console.log('The user rejected the request to connect with your dApp');
+      break;
+  }
+});
+</pre>
+
+
+
+====getBalance====
+Return balance of a specific asset for the given account.
+
+'''Method Interface'''
+
+<pre>
+function getBalance({ address: string, asset: string }): Promise<{ address: string, asset: string, amount: string }}>
+</pre>
+
+'''Input arguments'''
+
+address: string - The asset symbol the check the balance for the connected account
+
+asset: string - The asset symbol the check the balance for the connected account
+
+
+'''Success return value'''
+
+address: string - Address of the connected account
+
+asset: string - Asset symbol for the requested balance
+
+amount: string - Parsed balance of the requested asset
+
+
+'''Error return value'''
+
+string - `NO_PROVIDER`|`CONNECTION_DENIED`
+
+'''Example'''
+
+<pre>
+getBalance({
+  address: 'AeysVbKWiLSuSDhg7DTzUdDyYYKfgjojru',
+  asset: 'NKN'
+})
+.then((result: Result) => {
+  const {
+    address,
+    asset,
+    amount,
+  } = result;
+
+  console.log('Connected address: ' + address);
+  console.log('Asset: ' + asset);
+  console.log('Amount: ' + amount);
+})
+.catch((err: string) => {
+  switch(err) {
+    case NO_PROVIDER:
+      console.log('No provider available.');
+      break;
+    case CONNECTION_DENIED:
+      console.log('The user rejected the request to connect with your dApp');
+      break;
+  }
+});
+</pre>
+
+
+====send====
+Invoke a transfer of a specified amount of a given asset from the connected account to another account.
+
+'''Method Interface'''
+
+<pre>
+function send({ to: string, asset: string, amount: string, remark?: string }): Promise<txid: string>
+</pre>
+
+'''Input arguments'''
+
+to: string - Address of the receiver of the assets to be sent
+
+asset: string - The symbol of the asset to be sent
+
+amount: string - The parsed amount of the asset to be sent
+
+remark?: string - (Optional) Description of the transaction to be made
+
+
+'''Success return value'''
+
+txid: string - The transaction ID of the send invocation
+
+
+'''Error return value'''
+
+string - `NO_PROVIDER`|`CONNECTION_REFUSED`|`SEND_ERROR`|`MALFORMED_ADDRESS`|`CANCELED`
+
+'''Example'''
+
+<pre>
+send({
+  to: 'ATaWxfUAiBcQixNPq6TvKHEHPQum9bx79d',
+  asset: 'GAS',
+  amount: '0.0001',
+  remark: 'Hash puppy clothing purchase. Invoice#abc123',
+})
+.then((txid: string) => {
+  console.log('Send transaction success! Transaction ID: ' + txid);
+})
+.catch((err: string) => {
+  switch(err) {
```

Comment:
Few more errors to add:
1. ADDRESS_BLOCKED - The address you are trying to send is blocked/reported as a scam.
2. INSUFFICIENT_FUNDS - The user does not have enough funds for the amount specified (should show that for each asset)
3. INVALID_ASSETS - List of the invalid assets.

I'm also thinking if we should have a generic error that catches the other things... like passing wrong data types, etc.

Reply:
Nice, will add these two, and maybe a list of common error messages.

Reply:
I added `INSUFFICIENT_FUNDS`, but after thinking about `ADDRESS_BLOCKED`, I'm wondering if this is appropriate for this proposal. Since any list of known "bad addresses" would be custom compiled for each provider, and keeping a list of known scam addresses seems like a never ending battle. Any provider may be free to add these errors on their own beyond what is in this proposal if applicable.

Reply:
as for the generic error, I was wondering if it might be more valuable to simply return a raw error is it's not a known error. It may be more useful to help dapp developers to work with api providers to better understand what is happening.